include/tlapack/lapack/hetd2.hpp

```diff
@@ -0,0 +1,165 @@
+/// @file hetd2.hpp
+/// @author Skylar Johns, University of Colorado Denver, USA
+/// @author Weslley S Pereira, University of Colorado Denver, USA
+//
+// Copyright (c) 2021-2023, University of Colorado Denver. All rights reserved.
+//
+// This file is part of <T>LAPACK.
+// <T>LAPACK is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#ifndef TLAPACK_HETD2_HH
+#define TLAPACK_HETD2_HH
+
+#include "tlapack/base/utils.hpp"
+#include "tlapack/blas/axpy.hpp"
+#include "tlapack/blas/dot.hpp"
+#include "tlapack/blas/hemv.hpp"
+#include "tlapack/blas/her2.hpp"
+#include "tlapack/lapack/larfg.hpp"
+
+namespace tlapack {
+
+/** Reduce a hermitian matrix to real symmetric tridiagonal form.
+ *
+ * @return  0 if success
+ *
+ * @tparam uplo_t Either Uplo or any class that implements `operator Uplo()`.
+ *
+ * @param[in] uplo
+ *      - Uplo::Upper:   Upper triangle of A is referenced;
+ *      - Uplo::Lower:   Lower triangle of A is referenced;
+ *
+ * @param[in,out] A n-by-n symmetric matrix.
+ *      On exit, the main diagonal and offdiagonal contain the elements of the
+ * symmetric tridiagonal matrix B. The other positions are used to store
+ * elementary Householder reflectors.
+ *
+ * @param[out] tau Vector of length n-1.
+ *      The scalar factors of the elementary reflectors.
+ *
+ * @ingroup computational
+ */
+template <class matrix_t, class vector_t, class uplo_t>
+int hetd2(uplo_t uplo, matrix_t& A, vector_t& tau)
+{
+    using T = type_t<matrix_t>;
+    using real_t = real_type<T>;
+    using idx_t = size_type<matrix_t>;
+    using pair = pair<idx_t, idx_t>;
+
+    // constants
+    const idx_t n = ncols(A);
+    const real_t one(1);
+    const real_t zero(0);
+    const real_t half(0.5);
+
+    // check arguments
+    tlapack_check_false(uplo != Uplo::Lower && uplo != Uplo::Upper);
+    tlapack_check(nrows(A) == ncols(A));
+    tlapack_check((idx_t)size(tau) >= n - 1);
+
+    // quick return
+    if (n <= 0) return 0;
+
+    if (uplo == Uplo::Upper) {
+        //
+        // Reduce upper triangle of A
+        //
+
+        // Only access real part of the main diagonal
```

Comment:
A should be Hermitian, so the entry should already be real. This line seems more like an extra check to avoid problems, the comment should be changed to reflect that.

Reply:
We are trying to respect LAPACK where the convention is something like: "yes the matrix is Hermitian, yes diagonal entries are real, if a diagonal entry is not real, this is meant to mean "only use the real part and assume imaginary is intended to be zero but might not be set as such" of this entry." 
https://github.com/Reference-LAPACK/lapack/blob/7ac31faec5b5b02a2591414f30b9aed3005b32a8/BLAS/SRC/zherk.f#L131-L133